pages/en/lb4/Testing-Your-Extensions.md

```diff
@@ -0,0 +1,305 @@
+unit testing a controller for an application.
+
+To test a controller you want to instantiate a new instance of your controller
+class and test a function, providing a test double for constructor arguments as
+needed.
+
+**`src/controllers/ping.controller.ts`**
+```ts
+export class PingController {
+  @get('/ping')
+  ping(msg?: string) {
+    return `You pinged with ${msg}`;
+  }
+}
+```
+
+**`test/unit/controllers/ping.controller.ts`**
+```ts
+import {PingController} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('PingController() unit', () => {
+  it('.ping() (no arg)', () => {
+    const controller = new PingController();
+    const result = controller.ping();
+    expect(result).to.equal('You pinged with undefined');
+  });
+
+  it('.ping(\'hello\')', () => {
+    const controller = new PingController();
+    const result = controller.ping('hello');
+    expect(result).to.equal('You pinged with hello');
+  });
+});
+```
+
+You can find a more advanced example on testing controllers in [Unit test your Controllers](http://loopback.io/doc/en/lb4/Testing-your-application.html#unit-test-your-controllers).
+
+#### Decorators
+
+The recommended usage of a decorator is to store metadata about a class or a
+class method. The decorator implementation will usually also provide a function
+to retrieve the metadata stored by it based on the className and method name.
+So for a unit test, it is important to test that we can store metadata via a
+decorator and retrieve the correct metadata. *The retrieval gets tested as a
+consequence of validating the metadata was stored.*
+
+**`src/decorators/test.decorator.ts`**
+```ts
+export function test(file: string) {
+  return function(target: Object, methodName: string): void {
+    Reflector.defineMetadata(
+      'example.msg.decorator.metadata.key',
+      {file},
+      target,
+      methodName,
+    );
+  };
+}
+
+export function getTestMetadata(
+  controllerClass: Constructor<{}>,
+  methodName: string,
+): {file: string} {
+  return Reflector.getMetadata(
+    'example.msg.decorator.metadata.key',
+    controllerClass.prototype,
+    methodName,
+  );
+}
+```
+
+**`test/unit/decorators/test.decorator.ts`**
+```ts
+import {test, getTestMetadata} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('test.decorator (unit)', () => {
+  it('can store test name via a decorator', () => {
+    class TestClass {
+      @test('me.test.ts')
+      me() {}
+    }
+
+    const metadata = getTestMetadata(TestClass, 'me');
+    expect(metadata).to.be.a.Object();
+    expect(metadata.file).to.be.eql('me.test.ts');
+  });
+});
+```
+
+#### Mixins
+
+A Mixin is a function that extends a base Class (usually `Application` Class)
+and returns an anonymous class, adding new contructor properties, methods, etc.
+to the base Class. Due to the fact that the Mixin returns an anonymous class,
+it is not possible to write a unit test in isolation for a Mixin. The
```

Comment:
I would rephrase this to be more specific to TypeScript. You technically can write a unit test for a JavaScript mixin, since you can mix your artifacts into an empty class. Our mixins will most likely be constrained to classes that extend Application so it's not a *universal* truth, but it's still something important to keep in mind.